File: src/host/bridgeCore.ts
import { stat } from 'node:fs/promises';
import { resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { isSafeExternalUrl, type FileHit, type HostMsg, type WebviewHost, type WebviewMsg } from '@shared/protocol';
import type { Appearance } from '@shared/appearance';
import type { SessionManager, SessionViewer } from './SessionManager';
import type { SettingsCenter } from './settings';
import type { HostPlatform } from './platform';
import { msg } from './errors';
import { freezeHostMsg, HostMsgBatch } from './msgBatch';

export interface BridgeCoreDeps {
  manager: SessionManager;
  settings: SettingsCenter;
  platform: HostPlatform;
  appearance: () => Appearance;
}

export interface BridgeCoreOpts {
  // Where this view's HostMsgs go (the webview's postMessage, or the sidecar envelope for a viewId)
  post: (m: HostMsg) => void;
  host: WebviewHost;
  // How the view loads attachment blobs: `${blobBase}/${sessionId}/${blob}` (a webview URI of the sessions dir, or the shell's resource host)
  blobBase?: string;
  // The session it opens on: an id, the most recent session, or nothing → a fresh session
  initial?: string | { mostRecent: true };
}

// One per view (a VS Code webview, a JCEF browser): routes WebviewMsgs to its viewer / the manager / the settings center, resolves
// paths and makes the business decisions before asking the platform for an IDE action, and pushes changes back after coalescing.
// Nothing here knows how the view is rendered or transported
export class BridgeCore {
  readonly viewer: SessionViewer;
  private ready = false;
  private readonly batch = new HostMsgBatch(m => this.post(m));
  private readonly unsubscribe: (() => void)[] = [];

  constructor(private deps: BridgeCoreDeps, private opts: BridgeCoreOpts) {
    this.viewer = deps.manager.attach(opts.initial);
    this.unsubscribe.push(this.viewer.subscribe(ev => this.queue(ev)), deps.settings.subscribe(ev => this.queue(ev)));
  }

  // Failures are logged, never thrown at the transport: a bad message must not take the view down
  async handle(m: WebviewMsg): Promise<void> {
    try {
      await this.route(m);
    } catch (e) {
      this.deps.platform.log(`webview ${m.type} failed: ${msg(e)}`);
    }
  }

  private async route(m: WebviewMsg) {
    const { manager, platform } = this.deps;
    if (m.type === 'ready') {
      this.ready = true;
      // A view (re)opening is a cheap moment to re-check the executables (a change arrives as an `agents` event after init) and to pick up
      // sessions another window created since; the list is reconciled before ensureActive so a sidebar starting on "most recent" sees them
      void manager.reprobe();
      await manager.refreshIndex();
      await this.viewer.ensureActive();
      this.post({
        type: 'init',
        state: {
          host: this.opts.host, appearance: this.deps.appearance(), agents: manager.agents(), accounts: manager.accounts(), accountActions: manager.accountActions(), hidden: manager.hidden(),
          sessions: manager.sessions(), active: this.viewer.active(), blobBase: this.opts.blobBase,
          settings: this.deps.settings.view(), locale: this.deps.settings.locale(), home: platform.home(), cwd: platform.cwd(),
        },
      });
      return;
    }
    if (m.type === 'chatgptStatus') { this.post({ type: 'chatgptStatus', status: await manager.chatgptStatus() }); return; }
    if (m.type === 'openInEditor') { platform.openInEditor(m.sessionId ?? this.viewer.activeId); return; }
    if (m.type === 'openFile') {
      // Tool references are relative to the session they came from; a view showing another session ignores them
      const session = this.viewer.active();
      if (!session || session.id !== m.sessionId) return;
      try {
        const path = m.path.startsWith('file://') ? fileURLToPath(m.path) : resolve(session.cwd, m.path);
        const line = Number.isSafeInteger(m.line) && m.line! > 0 ? m.line : undefined;
        await platform.openResolvedFile(path, line);
      } catch (e) {
        platform.toast('error', msg(e));
      }
      return;
    }
    if (m.type === 'openPlan') {
      const plan = manager.planDocument(m.sessionId, m.planId);
      if (plan?.type === 'plan_document') {
        const exists = plan.path && await stat(plan.path).catch(() => undefined);
        await platform.openPlanDocument(exists && plan.path ? { path: plan.path } : { markdown: plan.markdown });
      }
      return;
    }
    if (m.type === 'openExternal') {
      if (isSafeExternalUrl(m.url)) platform.openExternal(m.url);
      else platform.log(`openExternal refused: scheme not on the allowlist (${m.url.slice(0, 80)})`);
      return;
    }
    if (m.type === 'searchFiles') {
      // Always answer, even on failure: the webview holds a promise per seq
      let files: FileHit[] = [];
      try { files = await platform.searchFiles(m.query); } catch (e) { platform.log(`searchFiles failed: ${msg(e)}`); }
      this.post({ type: 'files', seq: m.seq, files });
      return;
    }
    if (m.type === 'editTurn') {
      try {
        await manager.editTurn(m.edit);
        this.post({ type: 'editTurnResult', requestId: m.requestId });
      } catch (e) {
        this.post({ type: 'editTurnResult', requestId: m.requestId, error: msg(e) });
      }
      return;
    }
    if (await this.onSettingsMessage(m)) return;
    await this.viewer.handle(m);
  }

  // The settings page's requests; returns true when the message was its business
  private async onSettingsMessage(m: WebviewMsg): Promise<boolean> {
    const { manager, settings, platform } = this.deps;
    try {
      switch (m.type) {
        case 'setSetting': await settings.set(m.key, m.value); return true;
        case 'setAppearance': await settings.setAppearance(m.axis, m.value); return true;
        // A path from the inventory lists: files open in the editor, directories reveal in the OS file manager
        case 'openPath': {
          const s = await stat(m.path).catch(() => undefined);
          if (s?.isDirectory()) await platform.revealInOS(m.path); else await platform.openResolvedFile(m.path);
          return true;
        }
        case 'inventory': this.post({ type: 'inventory', agent: m.agent, inventory: await settings.inventory(m.agent) }); return true;
        case 'controls': {
          if (!m.fresh) { this.post({ type: 'controls', agent: m.agent, controls: await manager.knownControls(m.agent) }); return true; }
          // The probe's initialize brings the version back too; a fresh inventory after the list keeps the facts card in step
          this.post({ type: 'controls', agent: m.agent, controls: await manager.probeControls(m.agent) });
          this.post({ type: 'inventory', agent: m.agent, inventory: await settings.inventory(m.agent) });
          return true;
        }
        default: return false;
      }
    } catch (e) {
      platform.log(`settings ${m.type} failed: ${msg(e)}`);
      return true;
    }
  }

  // Streaming updates are dense; for the same message type within one batch window, keep only the latest
  private queue(m: HostMsg) {
    if (!this.ready) return;
    this.batch.push(m);
  }

  post(m: HostMsg) { this.opts.post(freezeHostMsg(m)); }

  pushAppearance() { this.post({ type: 'appearance', appearance: this.deps.appearance() }); }

  dispose() {
    this.batch.dispose();
    for (const u of this.unsubscribe) u();
    this.viewer.dispose();
  }
}

File: rust/crates/acpira-host/src/bridge_core.rs
//! One per view: routes WebviewMsgs to its viewer / the manager / the
//! settings center, resolves paths before asking the platform for an IDE action, and pushes changes back coalesced

use std::path::Path;
use std::sync::{Arc, Weak};
use std::time::Duration;

use serde_json::Value;

use acpira_shared::appearance::Appearance;
use acpira_shared::protocol::{FileHit, HostMsg, InitState, WebviewHost, WebviewMsg, is_safe_external_url};
use acpira_shared::settings::is_setting_key;
use acpira_shared::sidecar::{InitialView, PlanTarget};

use crate::acp::normalize::file_url_to_path;
use crate::i18n::tp;
use crate::session_manager::{SessionManager, Viewer};
use crate::settings::SettingsCenter;
use crate::sidecar::platform::SidecarPlatform;
use crate::store::data_dir::normalize;

/// Manager events within this window collapse to one post per message type
pub const BATCH_WINDOW: Duration = Duration::from_millis(30);

pub type Post = Arc<dyn Fn(HostMsg) + Send + Sync>;

/// Streaming updates are dense: within one batch window only the latest message per key survives, and an idle session edge
/// goes out at once together with whatever was pending
#[derive(Default)]
pub struct MsgBatch {
  pending: Vec<(String, HostMsg)>,
  armed: bool,
}

/// What a push asks of the owner
pub enum Pushed {
  /// Post these now
  Flush(Vec<HostMsg>),
  /// Start the window timer, then flush
  Arm,
  /// A timer is already running
  Wait,
}

impl MsgBatch {
  pub fn push(&mut self, m: HostMsg) -> Pushed {
    let idle = matches!(&m, HostMsg::Session { running: false, .. });
    let key = m.batch_key();
    match self.pending.iter().position(|(k, _)| *k == key) {
      Some(i) => self.pending[i].1 = m,
      None => self.pending.push((key, m)),
    }
    if idle {
      return Pushed::Flush(self.flush());
    }
    if self.armed {
      return Pushed::Wait;
    }
    self.armed = true;
    Pushed::Arm
  }

  pub fn flush(&mut self) -> Vec<HostMsg> {
    self.armed = false;
    std::mem::take(&mut self.pending).into_iter().map(|(_, m)| m).collect()
  }

  pub fn clear(&mut self) {
    self.pending.clear();
  }
}

pub struct BridgeCore {
  pub viewer: Arc<Viewer>,
  manager: Arc<SessionManager>,
  settings: Arc<SettingsCenter>,
  platform: Arc<SidecarPlatform>,
  appearance: Arc<dyn Fn() -> Appearance + Send + Sync>,
  post: Post,
  host: WebviewHost,
  blob_base: Option<String>,
  ready: std::sync::atomic::AtomicBool,
  batch: parking_lot::Mutex<MsgBatch>,
  settings_sub: parking_lot::Mutex<Option<u64>>,
  me: Weak<BridgeCore>,
}

impl BridgeCore {
  #[allow(clippy::too_many_arguments)]
  pub fn new(
    manager: Arc<SessionManager>,
    settings: Arc<SettingsCenter>,
    platform: Arc<SidecarPlatform>,
    appearance: Arc<dyn Fn() -> Appearance + Send + Sync>,
    post: Post,
    host: WebviewHost,
    blob_base: Option<String>,
    initial: Option<InitialView>,
  ) -> Arc<BridgeCore> {
    let viewer = manager.attach(initial);
    let core = Arc::new_cyclic(|me: &Weak<BridgeCore>| BridgeCore {
      viewer,
      manager,
      settings,
      platform,
      appearance,
      post,
      host,
      blob_base,
      ready: Default::default(),
      batch: Default::default(),
      settings_sub: Default::default(),
      me: me.clone(),
    });
    let weak = core.me.clone();
    core.viewer.subscribe(Arc::new(move |m| {
      if let Some(c) = weak.upgrade() {
        c.queue(m);
      }
    }));
    let weak = core.me.clone();
    let id = core.settings.subscribe(Arc::new(move |view, locale| {
      if let Some(c) = weak.upgrade() {
        c.queue(HostMsg::Settings { settings: view.clone(), locale });
      }
    }));
    *core.settings_sub.lock() = Some(id);
    core
  }

  /// Failures are logged, never thrown at the transport: a bad message must not take the view down
  pub async fn handle(self: Arc<Self>, raw: Value) {
    let kind = WebviewMsg::raw_type(&raw).to_owned();
    let m: WebviewMsg = match serde_json::from_value(raw) {
      Ok(m) => m,
      Err(e) => {
        self.platform.log(&format!("webview {kind} failed: {e}"));
        return;
      }
    };
    if let Err(e) = self.route(m).await {
      self.platform.log(&format!("webview {kind} failed: {e}"));
    }
  }

  async fn route(self: &Arc<Self>, m: WebviewMsg) -> anyhow::Result<()> {
    use WebviewMsg as W;
    let (manager, platform) = (&self.manager, &self.platform);
    match m {
      W::Ready => {
        self.ready.store(true, std::sync::atomic::Ordering::Release);
        // A view (re)opening is a cheap moment to re-check executables and pick up sessions another window created
        let m2 = manager.clone();
        tokio::spawn(async move { m2.reprobe().await });
        manager.refresh_index().await;
        manager.ensure_active_for(&self.viewer).await;
        let state = InitState {
          host: self.host,
          appearance: (self.appearance)(),
          agents: manager.agents(),
          accounts: manager.accounts(),
          account_actions: Some(manager.account_actions()),
          hidden: manager.hidden(),
          sessions: manager.sessions(),
          active: manager.view_of(self.viewer.active_id().as_deref()).map(|(raw, _)| raw),
          settings: self.settings.view(),
          locale: self.settings.locale(),
          home: platform.home(),
          cwd: platform.cwd(),
          blob_base: self.blob_base.clone(),
        };
        self.post_now(HostMsg::Init { state: Box::new(state) });
      }
      W::ChatgptStatus => {
        let status = manager.chatgpt_status().await;
        self.post_now(HostMsg::ChatgptStatus { status });
      }
      W::OpenInEditor { session_id } => platform.open_in_editor(session_id.or_else(|| self.viewer.active_id())),
      W::OpenFile { session_id, path, line } => {
        // Tool references are relative to the session they came from; a view showing another session ignores them
        let Some(active) = self.viewer.active_id().filter(|a| *a == session_id) else { return Ok(()) };
        let Some(cwd) = manager.session_cwd(&active) else { return Ok(()) };
        let resolved = if path.starts_with("file://") {
          file_url_to_path(&path).ok_or_else(|| anyhow::anyhow!("File URL must be absolute"))
        } else {
          Ok(normalize(&Path::new(&cwd).join(&path)).to_string_lossy().into_owned())
        };
        match resolved {
          Ok(p) => {
            if let Err(e) = platform.open_resolved_file(&p, line.filter(|l| *l > 0)).await {
              platform.toast("error", &e.to_string());
            }
          }
          Err(e) => platform.toast("error", &e.to_string()),
        }
      }
      W::OpenBlob { session_id, name } => {
        if let Some(p) = manager.blob_path(&session_id, &name)
          && let Err(e) = platform.open_resolved_file(&p.to_string_lossy(), None).await
        {
          platform.toast("error", &e.to_string());
        }
      }
      W::OpenPlan { session_id, plan_id } => {
        if let Some(plan) = manager.plan_document(&session_id, &plan_id) {
          let exists = match &plan.path {
            Some(p) => tokio::fs::metadata(p).await.is_ok(),
            None => false,
          };
          let target = match (exists, plan.path) {
            (true, Some(path)) => PlanTarget::Path { path },
            _ => PlanTarget::Markdown { markdown: plan.markdown },
          };
          platform.open_plan_document(target).await?;
        }
      }
      W::OpenExternal { url } => {
        if is_safe_external_url(&url) {
          platform.open_external(&url);
        } else {
          let head: String = url.chars().take(80).collect();
          platform.log(&format!("openExternal refused: scheme not on the allowlist ({head})"));
        }
      }
      W::SearchFiles { query, seq } => {
        // Always answer, even on failure: the webview holds a promise per seq
        let files: Vec<FileHit> = match platform.search_files(&query).await {
          Ok(f) => f,
          Err(e) => {
            platform.log(&format!("searchFiles failed: {e}"));
            vec![]
          }
        };
        self.post_now(HostMsg::Files { seq, files });
      }
      W::ExportSession { id, format } => match manager.export_session(&id, format).await {
        Ok(path) => {
          let p = path.to_string_lossy().into_owned();
          if let Err(e) = platform.open_resolved_file(&p, None).await {
            platform.toast("error", &e.to_string());
          } else {
            platform.toast("info", &tp("host.exported", &[("path", &p)]));
          }
        }
        Err(e) => platform.toast("error", &e.to_string()),
      },
      W::ListNativeSessions { agent } => {
        // Always answer: the popover is waiting on this agent's list
        match manager.list_native_sessions(&agent).await {
          Ok(sessions) => self.post_now(HostMsg::NativeSessions { agent, sessions, error: None }),
          Err(e) => self.post_now(HostMsg::NativeSessions { agent, sessions: vec![], error: Some(e.to_string()) }),
        }
      }
      W::EditTurn { request_id, edit } => {
        let error = manager.edit_turn(edit).await.err().map(|e| e.to_string());
        self.post_now(HostMsg::EditTurnResult { request_id, error });
      }
      W::SetSetting { key, value } => {
        if is_setting_key(&key)
          && let Err(e) = self.settings.set(&key, &value).await
        {
          platform.log(&format!("settings setSetting failed: {e}"));
        }
      }
      W::SetAppearance { axis, value } => {
        if let Err(e) = self.settings.set_appearance(&axis, &value).await {
          platform.log(&format!("settings setAppearance failed: {e}"));
        }
      }
      W::OpenPath { path } => {
        // A path from the inventory lists: files open in the editor, directories reveal in the OS file manager
        let dir = tokio::fs::metadata(&path).await.is_ok_and(|m| m.is_dir());
        let r = if dir { platform.reveal_in_os(&path).await } else { platform.open_resolved_file(&path, None).await };
        if let Err(e) = r {
          platform.log(&format!("settings openPath failed: {e}"));
        }
      }
      W::Inventory { agent } => {
        let inventory = self.settings.inventory(&agent).await;
        self.post_now(HostMsg::Inventory { agent, inventory });
      }
      W::Controls { agent, fresh } => {
        if fresh != Some(true) {
          let controls = manager.known_controls(&agent).await;
          self.post_now(HostMsg::Controls { agent, controls });
        } else {
          // The probe's initialize brings the version back too; a fresh inventory keeps the facts card in step
          let controls = manager.probe_controls(&agent).await;
          self.post_now(HostMsg::Controls { agent: agent.clone(), controls });
          let inventory = self.settings.inventory(&agent).await;
          self.post_now(HostMsg::Inventory { agent, inventory });
        }
      }
      other => manager.handle_for(&self.viewer, other).await,
    }
    Ok(())
  }

  fn queue(&self, m: HostMsg) {
    if !self.ready.load(std::sync::atomic::Ordering::Acquire) {
      return;
    }
    let pushed = self.batch.lock().push(m);
    match pushed {
      Pushed::Flush(queued) => queued.into_iter().for_each(|m| (self.post)(m)),
      Pushed::Arm => {
        let weak = self.me.clone();
        tokio::spawn(async move {
          tokio::time::sleep(BATCH_WINDOW).await;
          if let Some(c) = weak.upgrade() {
            c.flush();
          }
        });
      }
      Pushed::Wait => {}
    }
  }

  fn flush(&self) {
    let queued = self.batch.lock().flush();
    for m in queued {
      (self.post)(m);
    }
  }

  pub fn post_now(&self, m: HostMsg) {
    (self.post)(m);
  }

  pub fn push_appearance(&self) {
    self.post_now(HostMsg::Appearance { appearance: (self.appearance)() });
  }

  pub fn dispose(&self) {
    self.batch.lock().clear();
    if let Some(id) = self.settings_sub.lock().take() {
      self.settings.unsubscribe(id);
    }
    self.manager.detach(&self.viewer);
  }
}
